### PR TITLE
Rewrote parts of esil description

### PR DESCRIPTION
**Pull request by Lukas-Dresel**
Hopefully explains the concepts and env variables associated with ESIL a bit better.

### CODE REVIEW

**Comment by Maijin** [comment]
@Lukas-Dresel can you ask something about e asm.emuwrite and aeim while you're at it ?
Ressource:

https://twitter.com/radareorg/status/659917145451716609
and
https://twitter.com/radareorg/status/659679149385244673

**Comment by Lukas-Dresel** [comment]
I can try, but i don't really understand how it works, so i'm not sure. I haven't been able to use it yet, because i didn't know it required asm.emuwrite. I was wondering why the arguments never showed up.

Do you know why it requires asm.emuwrite? Doesn't seem like it would need to

**Comment by Maijin** [comment]
asm.emuwrite: Allow asm.emu to modify memory (WARNING)
just then use | aeim               initialize ESIL VM stack (aeim- remove) and should be gtg :) Try with latest git version on a simple crackme such as the IOLI linux 0x0 and you'll see the strcmp

**Comment by Maijin** [comment]
Like https://twitter.com/Maijin212/status/716677834559918080

**Comment by Lukas-Dresel** [comment]
I meant why does argument detection need to be able to modify memory? ;)

**Comment by Lukas-Dresel** [comment]
Because that's what it says in the tweet

**Comment by Lukas-Dresel** [comment]
But i'll definitely give it a shot

**Comment by radare** [comment]
because it requires emulation, and if you want proper emulation you need a writable stack, as well as a data sections. by default the emulator forbids any write for obvious reasons, but you can enable the io.cache which will not modify the real file or memory, but just a thin layer on top of it which i named “cache” and sounds quite obvious whats the purpose of it.

And yes, some implementations, or commands just enable the cache temporarly to emulate some stuff without the user noticing it at all. As long as this is a wip feature and its getting better over time its important to have full control over it. that’s why you can manage the entire esil vm from the r2 shell.

—pancake

**Comment by radare** [comment]
there are severl “argument detection” things here:

* determine which arguments are accessed by a function can be done statically
* determine which parameters are passed to a function sometimes requires emulation, and sometimes that emulation involves being able to emulate a stack
* symbol name can have an implicit function signature which suggests the argument description

also, in the first step where args are detected by static analysis, you can easily see how the stack grows (asm.stackptr) and where a specific variable is access for read or write

**Comment by Maijin** [comment]
@Lukas-Dresel ping

File: esil.md [hunk 1]
```diff
@@ -1,37 +1,50 @@
 # ESIL
 
-ESIL stands for 'Evaluable Strings Intermediate Language'. It aims to describe a Forth-like representation for every target CPU opcode semantics. ESIL representations can be evaluated (interpreted) in order to emulate individual instructions. Each command of an ESIL expression is separated by a comma. Its virtual machine can be described as this:
+ESIL stands for 'Evaluable Strings Intermediate Language'. It aims to describe a <a href="https://en.wikipedia.org/wiki/Forth_(programming_language)">Forth</a>-like representation for every target CPU opcode semantics. ESIL representations can be evaluated (interpreted) in order to emulate individual instructions. Each command of an ESIL expression is separated by a comma. Its virtual machine can be described as this:
```
Comment (crowell) [inline esil.md]: use md syntax for links https://daringfireball.net/projects/markdown/syntax#link

**Comment by Maijin** [comment]
@Lukas-Dresel

**Comment by radare** [comment]
@Maijin ok to merge?

**Comment by Maijin** [comment]
Yeah, but he didn't add the emuwrite and emu section.

**Comment by radare** [comment]
@Maijin feel free to add this info yourself

**Pull request merged**